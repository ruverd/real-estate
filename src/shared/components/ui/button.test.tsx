import { render, screen } from "@testing-library/react";
import { Mail } from "lucide-react";
import { describe, expect, it } from "vitest";

import { Button } from "./button";

const renderButton = (props?: React.ComponentProps<typeof Button>) => {
  return render(<Button {...props}>{props?.children || "Click me"}</Button>);
};

describe("Button", () => {
  it("should render with default variant", () => {
    renderButton();

    const button = screen.getByRole("button", { name: "Click me" });

    expect(button).toBeInTheDocument();
  });

  it("should render all variants correctly", () => {
    const variants = ["secondary", "outline", "ghost", "link"] as const;

    variants.forEach((variant) => {
      renderButton({ variant, children: `${variant} button` });

      const button = screen.getByRole("button", { name: `${variant} button` });

      expect(button).toBeInTheDocument();
    });
  });

  it("should render all sizes correctly", () => {
    const sizes = ["default", "sm", "lg", "icon"] as const;

    sizes.forEach((size) => {
      renderButton({ size, children: `${size} button` });

      const button = screen.getByRole("button", { name: `${size} button` });

      expect(button).toBeInTheDocument();
    });
  });

  it("should render as disabled", () => {
    renderButton({ disabled: true, children: "Disabled Button" });

    const button = screen.getByRole("button", { name: "Disabled Button" });

    expect(button).toBeDisabled();
  });

  it("should render with icon", () => {
    renderButton({
      children: (
        <>
          <Mail data-testid="mail-icon" />
          With Icon
        </>
      ),
    });

    const button = screen.getByRole("button", { name: "With Icon" });
    const icon = screen.getByTestId("mail-icon");

    expect(button).toContainElement(icon);
  });

  it("should apply custom className", () => {
    const customClass = "custom-class";

    renderButton({ className: customClass, children: "Custom Button" });

    const button = screen.getByRole("button", { name: "Custom Button" });

    expect(button).toHaveClass(customClass);
  });

  it("should render as child component when asChild is true", () => {
    renderButton({ asChild: true, children: <a href="#">Link Button</a> });

    expect(screen.getByRole("link")).toBeInTheDocument();
    expect(screen.queryByRole("button")).not.toBeInTheDocument();
  });

  it("should match snapshot with default variant", () => {
    const { container } = renderButton();

    expect(container.firstChild).toMatchSnapshot();
  });

  it("should match snapshot with icon and secondary variant", () => {
    const { container } = renderButton({
      variant: "secondary",
      children: (
        <>
          <Mail data-testid="mail-icon" />
          With Icon
        </>
      ),
    });

    expect(container.firstChild).toMatchSnapshot();
  });

  it("should match snapshot when disabled", () => {
    const { container } = renderButton({
      disabled: true,
      children: "Disabled Button",
    });

    expect(container.firstChild).toMatchSnapshot();
  });
});
