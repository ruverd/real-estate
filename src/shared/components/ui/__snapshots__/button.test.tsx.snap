// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button > should match snapshot when disabled 1`] = `
<button
  class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-[color,box-shadow] disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive cursor-pointer bg-primary text-primary-foreground shadow-xs hover:bg-primary/90 h-9 px-4 py-2 has-[>svg]:px-3"
  data-slot="button"
  disabled=""
>
  Disabled Button
</button>
`;

exports[`Button > should match snapshot with default variant 1`] = `
<button
  class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-[color,box-shadow] disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive cursor-pointer bg-primary text-primary-foreground shadow-xs hover:bg-primary/90 h-9 px-4 py-2 has-[>svg]:px-3"
  data-slot="button"
>
  Click me
</button>
`;

exports[`Button > should match snapshot with icon and secondary variant 1`] = `
<button
  class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-[color,box-shadow] disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive cursor-pointer bg-secondary text-secondary-foreground shadow-xs hover:bg-secondary/80 h-9 px-4 py-2 has-[>svg]:px-3"
  data-slot="button"
>
  <svg
    class="lucide lucide-mail"
    data-testid="mail-icon"
    fill="none"
    height="24"
    stroke="currentColor"
    stroke-linecap="round"
    stroke-linejoin="round"
    stroke-width="2"
    viewBox="0 0 24 24"
    width="24"
    xmlns="http://www.w3.org/2000/svg"
  >
    <rect
      height="16"
      rx="2"
      width="20"
      x="2"
      y="4"
    />
    <path
      d="m22 7-8.97 5.7a1.94 1.94 0 0 1-2.06 0L2 7"
    />
  </svg>
  With Icon
</button>
`;
